python/ray/serve/tests/test_metrics.py

```diff
@@ -526,6 +541,108 @@ def test_proxy_metrics_fields_not_found(serve_start_shutdown):
     print("serve_num_grpc_error_requests working as expected.")
 
 
+@pytest.mark.parametrize(
+    "serve_start_shutdown",
+    [
+        1,
+    ],
+    indirect=True,
+)
+def test_proxy_timeout_metrics(serve_start_shutdown):
+    """Test that HTTP timeout metrics are reported correctly."""
+
+    @serve.deployment
+    async def return_status_code_with_timeout(request: Request):
+        await asyncio.sleep(10)
+        return
+
+    serve.run(
+        return_status_code_with_timeout.bind(),
+        route_prefix="/status_code_timeout",
+        name="status_code_timeout",
+    )
+
+    r = requests.get("http://127.0.0.1:8000/status_code_timeout")
+    assert r.status_code == 408
+
+    # make grpc call
+    channel = grpc.insecure_channel("localhost:9000")
+    with pytest.raises(grpc.RpcError):
+        ping_grpc_call_method(channel=channel, app_name="status_code_timeout")
+
+    num_errors = get_metric_dictionaries("serve_num_http_error_requests")
+    assert len(num_errors) == 1
+    assert num_errors[0]["route"] == "/status_code_timeout"
+    assert num_errors[0]["error_code"] == "408"
+    assert num_errors[0]["method"] == "GET"
+    assert num_errors[0]["application"] == "status_code_timeout"
+
+    num_errors = get_metric_dictionaries("serve_num_grpc_error_requests")
+    assert len(num_errors) == 1
+    assert num_errors[0]["route"] == "status_code_timeout"
+    assert num_errors[0]["error_code"] == str(grpc.StatusCode.DEADLINE_EXCEEDED)
+    assert num_errors[0]["method"] == "/ray.serve.UserDefinedService/__call__"
+    assert num_errors[0]["application"] == "status_code_timeout"
+
+
+def test_proxy_disconnect_metrics(serve_start_shutdown):
+    """Test that disconnect metrics are reported correctly."""
+
+    @serve.deployment
+    async def return_status_code_with_disconnect(request: Request):
+        await asyncio.sleep(10)
+        return
+
+    serve.run(
+        return_status_code_with_disconnect.bind(),
+        route_prefix="/disconnect",
+        name="disconnect",
+    )
+
+    try:
+        requests.get("http://127.0.0.1:8000/disconnect", timeout=0.5)
+    except requests.exceptions.ReadTimeout:
+        pass
+
+    # make grpc call
+    channel = grpc.insecure_channel("localhost:9000")
+    stub = serve_pb2_grpc.UserDefinedServiceStub(channel)
+    request = serve_pb2.UserDefinedMessage(name="foo", num=30, foo="bar")
+    metadata = (("application", "disconnect"),)
+
+    def make_request():
+        try:
+            response = stub.__call__(
+                request, metadata=metadata
+            )  # Long-running RPC call
+            print("Response received:", response)
+        except grpc.RpcError as e:
+            print("Client disconnected:", e.code(), e.details())
+
+    thread = threading.Thread(target=make_request)
+    thread.start()
+
+    # Wait briefly, then forcefully close the channel
+    time.sleep(0.01)  # Give some time for the call to be initiated
```

Comment:
let's not rely on timing for tests unless absolutely necessary

you can use a `SignalActor` for this to make it deterministic (grep around other tests for similar patterns)

Reply:
good idea.